nixos/nginx: add recommended brotli settings

Izorkin opened the pull request:
###### Description of changes
Enable brotli module by default and add recommended brotli settings.

cc @ajs124 @trofi @aidalgol @erictapen
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Izorkin committed:
nginxModules.brotli: unstable-2020-04-23 -> unstable-2022-04-29

Izorkin committed:
nixos/nginx: add recommended brotli settings

Izorkin commented:
Needed add this patch https://patch-diff.githubusercontent.com/raw/google/ngx_brotli/pull/117.patch
I can't add it.

trofi reviewed (approved):
The change looks good `.nix`-code-wise. But I know nothing about `nginx` configs or idioms.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -22836,7 +22836,7 @@ with pkgs;
     # We don't use `with` statement here on purpose!
     # See https://github.com/NixOS/nixpkgs/pull/10474#discussion_r42369334
-    modules = [ nginxModules.rtmp nginxModules.dav nginxModules.moreheaders ];
+    modules = [ nginxModules.brotli nginxModules.rtmp nginxModules.dav nginxModules.moreheaders ];
```
Those changes where really dangerous for moreheaders because if you still had it in ``services.nginx.additionalModules`` then the module would be compiled twice in without any errors and only at runtime when handling they would crash. We must handle this better from now on.

My suggestions:
- assert if ``services.nginx.additionalModules`` combined with ``modules`` contains duplicates package and maybe pname wise and hard fail if so
- we must update the description of ``services.nginx.additionalModules`` because it still contain brotli.
and maybe we should only add it to ``services.nginx.additionalModules`` when brotli is enabled. Building nginx is really quick and fast.

Izorkin replied:
How to implement the first variant? I don't know how to implement module comparison mechanism.

SuperSandro2000 replied:
To be honest I am not sure but without it the migration path could become painful.

Izorkin replied:
In the release notes, I wrote that nginx is now built with brotli module by default. Maybe something needs to be added?

SuperSandro2000 replied:
I don't think a release note is enough. This can easily be missed and most people are not aware of it and could only be noticed when nginx is already crashing. Also from search.nixos.org it is also not obvious when adding new plugins that there are already some installed by default.

Izorkin replied:
Removed this changes.

SuperSandro2000 commented on pkgs/servers/http/nginx/modules.nix:
```diff
@@ -67,19 +67,14 @@ in
   brotli = {
-    src = let gitsrc = pkgs.fetchFromGitHub {
+    src = fetchFromGitHub {
       name = "brotli";
```
```suggestion
```

SuperSandro2000 commented on pkgs/servers/http/nginx/modules.nix:
```diff
@@ -67,19 +67,14 @@ in
-          --replace '$ngx_addon_dir/deps/brotli/c' ${lib.getDev pkgs.brotli}
-      '';
-    inputs = [ pkgs.brotli ];
+      fetchSubmodules = true;
```
```suggestion
```
we want to use our brotli if possible to not rely on commits like https://github.com/google/ngx_brotli/commit/6e975bcb015f62e1f303054897783355e2a877dc

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -140,6 +140,50 @@ let
       ''}
 
+      ${optionalString (cfg.recommendedBrotliSettings) ''
+        brotli on;
+        brotli_static on;
+        brotli_comp_level 5;
+        brotli_window 512k;
+        brotli_min_length 256;
+        brotli_types
```
Can you add a comment where we took those settings?

Izorkin replied:
Added comment.

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -140,6 +140,50 @@ let
+        brotli_window 512k;
+        brotli_min_length 256;
+        brotli_types
+          application/atom+xml
+          application/javascript
+          application/json
+          application/ld+json
+          application/manifest+json
+          application/rss+xml
+          application/vnd.geo+json
+          application/vnd.ms-fontobject
+          application/x-font-opentype
+          application/x-font-truetype
+          application/x-font-ttf
+          application/x-javascript
+          application/x-web-app-manifest+json
+          application/xhtml+xml
+          application/xml
+          font/eot
+          font/opentype
+          font/otf
+          font/truetype
+          image/bmp
+          image/svg+xml
+          image/vnd.microsoft.icon
+          image/x-icon
+          image/x-win-bitmap
+          text/cache-manifest
+          text/css
+          text/javascript
+          text/plain
+          text/vcard
+          text/vnd.rim.location.xloc
+          text/vtt
+          text/x-component
+          text/x-cross-domain-policy
+          text/xml;
```
I would recommend to dedupe this list with gzip and make them one.

Izorkin replied:
Gzip and brotli have different parameter names.

SuperSandro2000 replied:
I mean that we can move the mime types into a nix variable and re-use them for both compressions.

Izorkin replied:
Fixed. Removed duplicates mime types.
Updated PR.

Izorkin commented:
Moved update mailcap to this PR - https://github.com/NixOS/nixpkgs/pull/188537

Izorkin commented:
cc @Mic92 @ajs124 @SuperSandro2000

SuperSandro2000 commented:
I still think we must do something about https://github.com/NixOS/nixpkgs/pull/188002#discussion_r955861019

Izorkin commented:
> I still think we must do something about [#188002 (comment)](https://github.com/NixOS/nixpkgs/pull/188002#discussion_r955861019)

I did not find how to check the included modules for duplicates.

Can anyone help with this?

Izorkin commented:
Rebased PR and drop changes nginx modules.
Now we can merged PR?

cc @SuperSandro2000

RaitoBezarius commented:
Can we have a release note on this new flag which seems important for NGINX users?

SuperSandro2000 commented:
We still need to do something about the duplicate modules. I have ran into problems with that multiple times and nginx workers than just crash.

Izorkin commented:
Updated mime-types. Replaced:
 - `application/javascript` to `text/javascript`;
 - `application/rss+xml` to `application/x-rss+xml`;

These mime-types changes are present in the new `media-types` package - https://github.com/NixOS/nixpkgs/pull/203433



> We still need to do something about the duplicate modules. I have ran into problems with that multiple times and nginx workers than just crash.

In this PR I removed changes related to nginx modules.

Izorkin committed:
nixos/nginx: add release notes for `recommendedBrotliSettings`

RaitoBezarius commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -422,6 +469,14 @@ in
+        default = false;
+        type = types.bool;
+        description = lib.mdDoc ''
+          Enable recommended brotli settings.
```
Can you add here it is required to *add* the brotli module to `additionalModules` ? Otherwise people using nixos-search will get confused.

Izorkin replied:
Updated description.

ncfavier commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -41,4 +41,14 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - A new `virtualisation.rosetta` module was added to allow running `x86_64` binaries through [Rosetta](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment) inside virtualised NixOS guests on Apple silicon. This feature works by default with the [UTM](https://docs.getutm.app/) virtualisation [package](https://search.nixos.org/packages?channel=unstable& ...
 
+- A new parameter `recommendedBrotliSettings` has been added to `service.nginx`. Learn more about compression in Brotli format [here](https://github.com/google/ngx_brotli/blob/master/README.md)
+  To activate parameter need to build nginx with brotli module. For example:
```
```suggestion
- A new option `recommendedBrotliSettings` has been added to `service.nginx`. Learn more about compression in Brotli format [here](https://github.com/google/ngx_brotli/blob/master/README.md).
  To activate this option, you need to build nginx with the brotli module. For example:
```

Izorkin replied:
Fixed.

ncfavier commented on pkgs/servers/http/nginx/modules.nix:
```diff
@@ -67,15 +67,14 @@ in
     }; in
-      pkgs.runCommand "ngx_brotli-src" { } ''
-        cp -a ${gitsrc} $out
+      pkgs.runCommand "brotli" { } ''
+        cp -a ${src'} $out
         substituteInPlace $out/filter/config \
           --replace '$ngx_addon_dir/deps/brotli/c' ${lib.getDev pkgs.brotli}
       '';
```
Let's use `postFetch` for this.

Izorkin replied:
this variant not workin:
```
...
      rev = "6e975bcb015f62e1f303054897783355e2a877dc";
      sha256 = "sha256-nBstePpxUp58ebDpSlCWmhkCtiRlJsSDnagLkzxMZhE=";
      postFetch = ''
        substituteInPlace $out/filter/config \
          --replace '$ngx_addon_dir/deps/brotli/c' ${lib.getDev pkgs.brotli}
      '';
    };
    inputs = [ pkgs.brotli ];
  };
```
Build log:
```
nginx> checking for getaddrinfo() ... found
nginx> configuring additional modules
nginx> adding module in /nix/store/2ysp5ichpccf4lv1wp2qcwz0bmm840f1-rtmp
nginx>  + ngx_rtmp_module was configured
nginx> adding module in /nix/store/6pb7j6kymf3y4xs5blp3g8mwin2j22kk-dav
nginx>  + ngx_http_dav_ext_module was configured
nginx> adding module in /nix/store/zx05c5lb0ivmjx6fmaa5csjrqpc7fwjc-moreheaders
nginx>  + ngx_http_headers_more_filter_module was configured
nginx> adding module in /nix/store/7d28d91dmzwpf4ahm0wjjy88ajqrwc8c-brotli
nginx> ./configure: error: no /nix/store/7d28d91dmzwpf4ahm0wjjy88ajqrwc8c-brotli/config was found
error: builder for '/nix/store/0aia49f14xb35gqzz9imb7hmviyc2dyr-nginx-1.22.1.drv' failed with exit code 1;

```

ncfavier replied:
Weird, it looks like whenever the `fetchFromGitHub` derivation depends on another derivation (here `pkgs.brotli`) the output path at evaluation time doesn't match the built output path:

```
$ nix-instantiate --eval --expr 'with import ./. {}; nginxModules.brotli.src.outPath'
"/nix/store/b81gnvm5f4b5xj4lc2dcif2syb7693if-brotli"
$ nix-build --expr 'with import ./. {}; nginxModules.brotli.src'
/nix/store/dsfqivq4ch12hm1impylix7dnxxlv6gr-brotli
```

Let's leave it be then

Izorkin replied:
> Let's leave it be then

Ok.

ncfavier replied:
Reported here https://github.com/NixOS/nix/issues/7150#issuecomment-1351889450

ncfavier commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -422,6 +469,15 @@ in
+        default = false;
+        type = types.bool;
+        description = lib.mdDoc ''
+          Enable recommended brotli settings. Required add the brotli module to
+          services.nginx.additionalModules.
```
```suggestion
          Enable recommended brotli settings. Learn more about compression in Brotli format [here](https://github.com/google/ngx_brotli/blob/master/README.md).
          
          Requires adding the brotli module to `services.nginx.additionalModules`:
          
          ```nix
          services.nginx = {
            additionalModules = [ pkgs.nginxModules.brotli ];
            recommendedBrotliSettings = true;
          };
          ```
```

Izorkin replied:
Thanks, fixed.

Izorkin commented:
Can be merge PR?

ncfavier commented:
> Implemented duplicate module detection in #206061. If that lands we can add brotli to `additionalModules`.

I'd like to wait for this so we don't have to go back and change the documentation and release notes later.

Izorkin commented:
Resolve conflicts.

Izorkin commented:
> Could you do this please?

It's going to take some time. I don't have these changes saved anywhere :(

Izorkin commented:
If it works out, I'll have it done by tonight.

Izorkin commented:
Or do it in a separate PR?

ncfavier commented:
I've done it myself

@ofborg test nginx discourse

Izorkin commented:
Hmm, these changes are not required?
```diff
-    modules = [ nginxModules.rtmp nginxModules.dav nginxModules.moreheaders ];
+    modules = [ nginxModules.brotli nginxModules.rtmp nginxModules.dav nginxModules.moreheaders ];
```
This module included to default:
```
        example = literalExpression "[ pkgs.nginxModules.moreheaders ]";
```

ncfavier commented:
I don't think we want to add brotli by default since that will cause an error now if the module also includes it?

Good point about the example though, I'll change it

Izorkin commented:
> I don't think we want to add brotli by default since that will cause an error now if the module also includes it?

I think it should be part of the basic functionality.

ncfavier commented:
In that case we could ignore duplicates instead of throwing, if they have the same source. That can be done in a follow-up PR though since it wouldn't change the module's interface.

Izorkin commented:
> In that case we could ignore duplicates instead of throwing

Do not needed ignore duplicate modules. If nginx is built with redundant modules, it will work unpredictably.

I'll try it in another PR.

ncfavier commented:
Sorry, by 'ignore' I mean only keep the first one, silently discarding the duplicates.

Izorkin commented:
Resolve conflicts.

Izorkin commented:
Resolve conflicts.

RaitoBezarius commented:
@ofborg test nginx

Izorkin commented:
Can be merge PR?

necauqua commented:
`compressMimeTypes` seems to have `application/javascript` missing, which is a shame since a lot of text is javascript :)

And `compressMimeTypes` could've been an option too (to fix/adjust things like these) - for now to fix this for me it seems that I have to override gzip_types and the like manually

RaitoBezarius commented:
> Per [IETF RFC 9239](https://www.ietf.org/rfc/rfc9239.pdf) text/javascript is now standard and application/javascript is now considered obsolete.

Izorkin commented:
Now `application/javascript` has been moved to `text/javascript`.
You can define new mime types via the parameter `services.nginx.defaultMimeTypes`. Example:
```
  services.nginx.defaultMimeTypes = "/etc/nginx/mime.types";
```
As a sample mime.types can be taken from here - https://github.com/NixOS/nixpkgs/pull/203433 .

necauqua commented:
> Now `application/javascript` has been moved to `text/javascript`.

> > Per [IETF RFC 9239](https://www.ietf.org/rfc/rfc9239.pdf) text/javascript is now standard and application/javascript is now considered obsolete.

Well all I'm saying is that the (more or less) default nix nginx configuration currently does not compress .js files if you enable the recommended compression setting - as nginx (by default with the current nix setup) sees .js files as having the `application/javascript` mime type.

Guessing that to comply with this obsoletion from the linked RFC nginx should now recognize (and serve in the Content-Type header) this new `text/javascript` mime type for js files, but without extra hoops it does not 🤷

Ideally for backwards compatibility and whatnot I'd personally just leave `application/javascript` in that list as it doesn't hurt in any way, but that's whatever.

Izorkin commented:
Once the `media-types` package can be added, it will be possible to correct this error
